A general Masked class for Quantity and other ndarray subclasses

mhvk opened the pull request:
Split off from #10423, to include only regular arrays and Quantity and its subclasses.  From #10423, the general comments still hold:

It may be worth pointing out two specific things:
- I very much on purpose decided to treat reductions different from `MaskedArray` - any time there is any items in the reduction masked, the result is masked as well. So, e.g., for `.sum()`, masked elements are not implicitly replaced by zero. The logic is that one cannot interpret the sum meaningfully without knowing how many elements actually contributed. Conversely, following that same logic, for methods where skipping a masked element does make sense (like `.mean()`, where one divides by the number of unmasked elements), I do use it.
- ~Currently, the implementation does not support structured arrays. The question for those, which I haven't really found a "best" answer for, is whether the mask should be per full item, or whether the mask itself should also be a structured array, i.e., allow masking individual pieces. `MaskedArray` does the latter and it is quite a hassle to keep track, but perhaps better.~  EDIT: followed `MaskedArray` and have a structured mask.

This is definitely at a point where some further design choices have to be made, and it would be good to step back a bit and decide what exactly we want masking to mean. 

Part of the difficulty is that it cannot immediately used for `MaskedColumn`, since some design choices which are different from `MaskedArray`, so it would need a kind of compatibility layer if we want to keep exact one-to-one matches, and/or some way to decide which type one wants.

fixes #1852, #1857, #7367, #10792

mhvk commented:
@taldcroft - I added brief documentation as well as more comments and docstrings to the code. I still need to think a bit about the auto-generation with `Masked(UnmaskedClass)`...

taldcroft commented on docs/utils/masked.rst:
```diff
@@ -0,0 +1,136 @@
+well defined, such as for getting the mean and other sample properties such as
+the variance and standard deviation.
+
+A third difference is more conceptual.  For `~numpy.ma.MaskedArray`, the
```
I don't totally understand this paragraph (and I actually wrote an ndarray sublcass once, albeit long ago). I think a more useful discussion for most users to just note that while `mq = np.ma.MaskedArray([1, 2]*u.m)` appears to do something useful on first glance, it actually just does not work. E.g.
```
In [40]: np.ma.array([1,2]*u.m) / u.s                                           
Out[40]: <Quantity [1., 2.] 1 / s>
```
So it is not just a conceptual break, it is paradigm change that allows usefully wrapping ndarray-subclasses.

You might also note here that the astropy `Column` (which is an ndarray subclass) can easily be wrapped with `Masked` and it all *just works*. (But that we choose not to do so because of the behavior changes).

mhvk replied:
Yes to examples!  I added some and also mention `Column`.

taldcroft reviewed (commented):
@mhvk - thanks, the docs are a good place to start some API discussions, and a gentle introduction to reviewing this.  :-)

taldcroft commented:
Question - will this be able to support setting the mask via `__setitem__` in a way like `mq[5] = np.ma.masked`? I like this idiom better than `mq.mask[5] = True`.

taldcroft commented:
Question 2 - what's your sense of how much it would take to apply this to a non-ndarray class like `SkyCoord` (and even `Time` if `Masked` really works out well).  I'm not especially attached to the `Nan` in `jd2` implementation, except that it basically works.

mhvk commented on docs/utils/masked.rst:
```diff
@@ -0,0 +1,136 @@
+should be substituted for any masked values, with
+:meth:`~astropy.utils.masked.Masked.unmask`::
+
+  >>> mq.unmasked
```
Yes, I tried to have the (very limited) API to have "mask" in the names to avoid overloading. Here, I particularly liked that I could then have a matching `unmask()` method.

But as you will see from the initializer, "data" and "mask" is definitely in my mind too...

mhvk commented on docs/utils/masked.rst:
```diff
@@ -0,0 +1,136 @@
+dividing a length by a time: if the unit of a masked quantity is changing, why
+should its value not change?).  But it also helps to keep the implementation
+considerably simpler, as the |Masked| class now primarily has to deal with
+propagating the mask rather than deciding what to do with values.
```
Indeed. I thought about this quite a bit, but really disliked how `MaskedArray` specifically has to check every value for being very close to 0 for division, not negative for square root, etc.

I have thought a bit about the option to have different kinds of `Mask` classes (which, e.g., could also deal with `.sum()` differently), and slightly tried to keep a path open for that.  For now, though, it seemed most important to just get mask propagation work properly.

mhvk commented on docs/utils/masked.rst:
```diff
@@ -0,0 +1,136 @@
+
+  >>> mq.unmasked
+  <Quantity [1., 2., 3.] m>
+  >>> mq.unmask(-75*u.cm)
```
See above, but I do call the argument `fill_value`, and the example should at least be clearer if I use that.

mhvk commented:
Thanks, very useful!

On your questions:
> will this be able to support setting the mask via `__setitem__` in a way like `mq[5] = np.ma.masked`? I like this idiom better than `mq.mask[5] = True`.

Yes, it does (I like it too!). It is why `__setitem__` looks slightly odd (`_data_mask` returns `None, True` for `np.ma.masked`): 
```
    def __setitem__(self, item, value):
        value, mask = self._data_mask(value)
        if value is not None:
            self.unmasked[item] = value
        self.mask[item] = mask
```

> what's your sense of how much it would take to apply this to a non-ndarray class like `SkyCoord` (and even `Time` if `Masked` really works out well).  I'm not especially attached to the `Nan` in `jd2` implementation, except that it basically works.

I think it should not be too hard, mostly because in the end it is quite similar to using `nan` - just need something that collects and sets the masks. I am hopeful that it will need just a single `MaskedShapedLikeNDArray` class, with everything that depends on `ShapedLikeNDArray` then automatically able to use it. 

But it may still take some effort as one does need to ensure that masks are shared or otherwise kept consistent between attributes, and verifying that there are no functions or ufuncs required for the transformations which the current code doesn't deal properly with.

Aside: something I have spent too much time on (but is fun) is to get this to work reliably:
![image](https://user-images.githubusercontent.com/2789820/101406300-4435c300-38a7-11eb-8555-aed38815330e.png)

But it is different between an ANSI-emulating console and, say, a jupyter notebook in a browser:
![image](https://user-images.githubusercontent.com/2789820/101406118-0e90da00-38a7-11eb-932d-c13f64861ac6.png)

mhvk commented:
@taldcroft - I updated the documentation page with your suggestions, and also made two changes that you had suggested elsewhere:
* Ensure that, say, `MaskedQuantity` can be initialized just like a `Quantity`, but with a `mask` argument (keyword only);
* Allow `Masked(Quantity)`, etc., as a short-cut to create a "default" masked class.

Both were fairly easy and will extend without problem to container classes like `SkyCoord` (once those work in the first place, of course...). So, that gives some more confidence that this route is actually a good one!

I'm trying to think what point is best to stop for this PR.  One option would be about where it is, i.e., with a PR that does not yet touch anything else.

Another option might be to see to what extent the new `MaskedQuantity` can be used in a `QTable` wherever now `MaskedColumn` is used in a `Table`.

I'm also interested in experimenting with `Time`, which might well be easier that representations and coordinates because of the infrastructure you made there already, but that is almost certainly much better done in follow-up....

mhvk commented:
Just for amusement, some things already work better than I had thought possible:
```
In [1]: from astropy.utils.masked import Masked

In [2]: from astropy.coordinates import SkyCoord

In [3]: sc = SkyCoord([0., 10.], [20., 30.], unit='deg')

In [4]: msc = sc._apply(Masked, mask=[True, False])

In [5]: msc.ra, msc.dec
Out[5]: (<MaskedLongitude [———, 10.0] deg>, <MaskedLatitude [ ———, 30.0] deg>)

In [6]: msc
Out[6]: 
<SkyCoord (ICRS): (ra, dec) in deg
    [( 0., 20.), (10., 30.)]>
```
Latter is because display somehow loses the mask...

Also, propagation will need better support of gufuncs:
```
In [16]: msc.cartesian
TypeError: operand type(s) all returned NotImplemented from __array_ufunc__(<ufunc 's2c'>, '__call__', <MaskedLongitude [———, 10.0] deg>, <MaskedLatitude [ ———, 30.0] deg>, out=(<Quantity [[ 0., 20.,  1.],
           [10., 30.,  1.]]>,)): 'MaskedLongitude', 'MaskedLatitude', 'Quantity'
```

taldcroft commented:
@mhvk - amazing progress. I need to find a block of time to give this a deeper review, but to be honest I expect that to be more functional than deep line-by-line of all the new stuff. Overall I agree this may be a decent waypoint to get this into master in order to facilitate subsequent smaller, more manageable, PRs.

Just one comment which may disappoint you, but I'm not a huge fan of the strikethrough. There is this unresolvable philosophical schism of whether the data are "masked" or "invalid/missing". From the database perspective it is typically the latter. If you do a table join and there are missing elements, the data behind the mask is junk and should not be displayed. If you want a configuration option to use strikethrough that would be OK, but overall I'd prefer to not rock the boat quite so much and just stick with `--` for missing / masked data.

mhvk commented:
OK, sounds good!  Right now, in a separate branch where I've been trying, it is definitely only an option to use strikethrough - indeed, beyond your good point that usually you don't actually want to see what is behind the mask, it seems sphinx/.rst does not support strikethrough so it will be difficult to document it!

I don't think it needs to hold up the PR, but what I still do hope to ensure is that all erfa ufunc work (right now some of the gufunc among them do not - one needs to infer what axis is being acted on so one act on the mask similarly; not super-difficult, but a bit tricky and tedious). I'm really hopeful, though, that the follow-up PR for containers like `SkyCoord` will turn out to be quite easy!

mhvk commented:
Just made a few further updates to ensure that representations, coordinates and times can propagate masks if they are present (for `Time`, needs a bit of a hack to initialize).

taldcroft commented:
@mhvk - just to let you know that I've been a bit slammed by day-job stuff but this is still on my radar. Holidays are coming!

mhvk commented:
@taldcroft - no worries if this takes some time... it is a very big PR.

In the meantime, I added support for structured dtype, since that is needed for `Time`.

I also have branch that uses this internally in `Time`, which works well (except I still have to get saving to files to work). One question: In principle, I could make it such that if you input `np.ma.MaskedArray`, you also get that class out for all `.to_value()`, etc. Is this worth doing? I worry it will break more than it helps, since `np.ma.MaskedArray` does not work for quite a few cases (e.g., all `erfa` ufuncs...). My tendency would be to accept it on input, but always use `Masked` inside.

mhvk commented:
Rebased and updated with substantially increased coverage of numpy functions. Somewhat better documentation of functions where behaviour is non-obvious.

taldcroft commented:
> My tendency would be to accept it (np.ma.MaskedArray) on input, but always use Masked inside.

Agreed. I think it would be confusing / fragile to do otherwise.

mhvk commented:
Just checked: with #11222 `Masked` array and quantity round-trip to fits using `serialization_method='null_value'`.

saimn commented:
How does this relate to the plans on Numpy side for a new masked array class/module ? At first glance it seems that `MaskedNDArray` and the mechanism to get a masked class for ndarray subclasses could have a wider usage and be useful outside of Astropy ?

mhvk commented:
@saimn - I have followed the numpy discussion somewhat and while there is a prototype, some initial trials made clear that it was too similar to `np.ma.MaskedArray` and didn't really work with `Quantity` (my main target, although really I hope even more that I can get masked Time and SkyCoord "for free"). The approach here is somewhat radical, with inserting the masked class "below" the main class, at some level making it an `ndarray` substitute as far as `Quantity` is concerned. 

I do indeed hope that if we can prove this approach is useful and works well, it will be taken up by others. Indeed, I've tried fairly hard to ensure I do not bake in any astropy dependencies...

saimn commented:
@mhvk - Ok! But given the significant number of lines of code, I guess one question is what would happen if Numpy were to adopt another approach, possibly incompatible ? Would it make sense to publish the core (non-Astropy) part independently and discuss this with the Numpy community ?

mhvk commented:
The numpy people seem to feel quite strongly one should not subclass `ndarray` in the first place, e.g., arguing that for `Quantity`, the unit should really be thought of as part of the `dtype` (an argument which I agree with in principle but it would have meant not having had `Quantity` still!). They are also rather hesitant to adding complex new things to numpy itself, feeling that `np.ma`, like `np.matrix`, would ideally live outside of numpy (again, quite logically). I also foresee that they would want more compatibility with `np.ma`, while I really prefer my more minimal design that avoids some of the mistakes which I think have become obvious in `np.ma` (in hindsight!). I'm still happy to discuss with numpy (just like I think units should really move up to numpy or scipy), but feel we may as well start in astropy...

Note that while it is quite a bit of code, one thing going in favour of having it here is that the same approach of inserting oneself in the class hierarchy is used in `uncertainty.Distribution` (indeed, having seen how implementing that approach made `Distribution` quite robust gave me the idea that it should work for `Masked` too).

mhvk commented:
p.s. I can see the case for a separate package initially, though one of the real goals here is to finally allow masked `Quantity` (and thus `SkyCoord`) in tables, so presumably this would be a hard dependency for astropy anyway. Hence it seems easier to work within astropy... In particular, I'd really like to get to the point where I can make smaller follow-up PRs - I know this one is almost impossible to review... (suggestions for splitting up most welcome...)

taldcroft commented:
@mhvk - a key goal I have for this week is to get this reviewed and ready to be merged. Being honest, the bulk of the new masked array code is beyond what I can do for a line-by-line review, but since it is experimental I don't see a problem with merging subject to these review criteria:

- [ ] Complete test coverage. Looks mostly there but codecov does note a handful of uncovered code paths.
- [ ] Careful independent review of parts of the code that interact with or impact existing astropy functionality. I will be focusing on this.
- [ ] Review of documentation.

My plan is to look at the documentation and new tests as an introduction. In the meantime I wonder if a rebase might be a good thing? I was also thinking about you sending mail to astropy-dev with an invitation to have a look and review. Perhaps there are others that might have the bandwidth to dig into the new core code.

mhvk commented:
@taldcroft - thanks! I had been planning to wait with a rebase until a few off-shoots were merged (which help keep things a bit simpler here...), but #11210 is still waiting, so I'll just go ahead and rebase now -- I can deal with possible merge conflicts when the time comes.

I hope things are more or less clear...  I wish there was a clearer way of how to deal with large code changes like these. In hindsight I perhaps should have written an APE, but I fear then nothing would ever have gotten done - as it is, it is not something I need all that much myself, but more a technically interesting challenge to get better masked support (if only so that quantities in tables become feature-compatible with columns!).

Anyway, I think your approach makes sense; indeed, following your earlier advice, I tried to minimize impact on the rest of astropy - mostly, it is just small changes/oversights which make things work.

taldcroft commented on docs/utils/masked/index.rst:
```diff
@@ -0,0 +1,165 @@
+should be substituted for any masked values, with
+:meth:`~astropy.utils.masked.Masked.unmask`::
+
+  >>> mq.unmasked
```
Coming back to the name of this attribute, I suspect I'm not the only one that will yearn for the `np.ma` consistency of calling it `data`. In particular I do find this a bit surprising:
```
In [23]: ma = Masked(data=[1,2], mask=[True, False])

In [24]: ma.data
Out[24]: <memory at 0x7ff37059a940>
```
I understand this is just what `np.ndarray` is exposing, but still, the `data` / `mask` duo is easy to explain and has a long heritage.

I suppose there is a question of whether to purposely diverge from the np.ma.MaskedArray API given the behavior differences.

mhvk replied:
It was indeed a very conscious choice to make the new attributes and methods extremely limited and ensure that they always include "mask" in their name - the fact that `Column.data` exists and the resulting problems for `MaskedColumn` are a prime inspiration!

But I think the argument becomes clearer if you look at other classes. E.g., for an ndarray subclass like `MaskedAngle`, does it really makes sense for `angle.data` to be an `Angle`? `angle.unmasked` would be more informative.

And for container classes like `Time` and `SkyCoord`, I'd hope to give those attributes and methods with similar names (following what you did already in `Time`), i.e., they should gain at least `.mask` and perhaps also `.unmask()` and `.unmasked`, which will be clearer than `.data` (in fact, it cannot be `.data` for `SkyCoord`, as that already returns the representation - surely, we'd want it to return the masked representation for a masked SkyCoord).

taldcroft replied:
Fair enough!

taldcroft commented on docs/utils/masked/index.rst:
```diff
@@ -0,0 +1,165 @@
+
+  >>> mq.unmasked
+  <Quantity [1., 2., 3.] m>
+  >>> mq.unmask(fill_value=-75*u.cm)
```
Again, I would vote to call this `mq.filled(fill_value=-75*u.cm)`. This way users / developers don't need to know if a column is a MaskedColumn or a MaskedQuantity in order to fill it.

That same argument applies to the `data` attribute. Existing code which assumes the data under the mask in an astropy Table masked column is in `data` will just work with `MaskedQuantity`.

mhvk replied:
This one I don't feel as strongly about - it is mostly that `unmasked = property(unmask)` just seems nice. But `.filled()` tells the reader more clearly what is happening if you give the value as a positional argument. I'm also less worried about name clashes.

mhvk replied:
p.s. And in the current implementation it isn't even true anymore that `unmasked=property(unmask)`...

Still, thinking ahead to a `SkyCoord`, is `.filled()` logical unless you are already thinking that it is a masked class? (The idea would be that the method would always exist, and if no mask is present, would just return the instance itself; I think `sc.unmask(fill_value=...)` would be clearer.

But then again, I've had quite a few times when writing tests that I wrote `...filled()`...

mhvk replied:
OK, since I even called all my tests "filled", let's just go with the familiar! Done in the latest push.

taldcroft reviewed (commented):
First go at high-level API. Overall this is really amazing work.

mhvk commented on docs/utils/masked/index.rst:
```diff
@@ -0,0 +1,165 @@
+to marking bad elements in arrays with NaN (not-a-number).  Like those NaN,
+the mask just propagates, except that for some operations like taking the mean
+the equivalence of `~numpy.nanmean` is used.
+
```
Hmm, that example is I fear an immediate oversight, since
```
ma.min()
# 2
```
and surely `argmin()` and `min()` should be consistent!

And, yes, it is a really silly bug: I actually do mean to ignore all masked elements and to do that replace those with the maximum in the array, which I thought was good enough, but in this particular example, it ends up just doing `np.array([2, 2]).argmin()` which yields 0. Oops...

Bottom line is that clearly I should have the list you mentioned!

mhvk replied:
On providing analogues to the `nan*` functions: I actually considered just overriding those functions to act on `ma.unmask(fill_value=np.nan)` (i.e., take masked elements to be `nan`, which needs some thought for integer dtype, obviously...); indeed, I have an extra commit that just does that but without any tests.

taldcroft replied:
I thought this infrastructure was meant to work on anything that could be in an array, e.g. strings or bool or whatever.

mhvk replied:
Indeed, it should!  Which is probably one reason I didn't continue with the `nan*` overrides...

p.s. Fix for `argmin()`, `argmax()` coming up... Right now, like for `np.ma.MaskedArray`, I have it return `0` if all elements are masked. Should be a masked `0` instead?

mhvk replied:
p.p.s. While it should be able to handle all dtype, I should be clear that the tests are almost all for float - plus a few for structured arrays consisting of floats - I re-use many for `Quantity` and `Longitude`...

taldcroft replied:
> Fix for argmin(), argmax() coming up... Right now, like for np.ma.MaskedArray, I have it return 0 if all elements are masked. Should be a masked 0 instead?

I think that case should be an exception like `np.argmin([])`. Returning 0 is definitely the wrong answer.
```
In [1]: np.argmin([])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: attempt to get argmin of an empty sequence
```

taldcroft replied:
About the `nan*` functions, having analogs for those would be useful I think and also provide a bridge for migrating from `np.ma.MaskedArray`. They can certainly fail for non-applicable types (e.g. str) just as numpy now does. `np.nanmin(['a', 'b'])` gives a `TypeError: cannot perform reduce with flexible type`.

I guess this is not a hard requirement for this PR but I would strongly suggest as a followup.

mhvk replied:
For the `nan*` functions, I was concretely suggesting to *override* those to mean any masked or nan elements are skipped. This as I don't really see those functions being very useful otherwise in the presence of masked values. But given the length of this PR and the thought required about how to implement it for integers, etc., perhaps best done as follow-up.

mhvk replied:
@taldcroft - I realize we never resolved what `argmin/max` should do for the case that everything is masked. You suggested raising an error, but one problem with that is what should happen if one does `np.argmin(ma, axis=0)` and only some slices along axis 0 are fully masked? To me, treating it a bit more as "sort all values, with masked at the end, and then return the index of the value that is now at 0" is perhaps the best after all. 

But in principle, one could also return a masked index array.

mhvk replied:
p.s. Added this to tracking issue #11529

mhvk replied:
p.p.s. Also added the nanfunctions to #11529

taldcroft replied:
> what should happen if one does np.argmin(ma, axis=0) and only some slices along axis 0 are fully masked? 

I don't really see the `axis=0` case as being any different. The values you get out should be exactly the same if you do a Python loop over one axis of a 2-d array and compute argmin for each 1-d subset vs. using a single vectorized call.

Masking the value does not help in the current implementation, and in fact seems to allow silently getting the wrong answer:
```
In [12]: idx = Masked(0, mask=True)  
In [13]: idx
Out[13]: MaskedNDArray(—)
In [14]: np.arange(5)[idx]
Out[14]: 0  # Yikes, I expected an exception!
```

> "sort all values, with masked at the end, and then return the index of the value that is now at 0"

That seems like a weird ad hoc definition that fits the current code behavior, but nobody would really expect that. If you have no (valid) values in an array and ask for the index of the minimum, I cannot see any outcome besides an exception, or maybe a masked scalar, being reasonable. 

Returning a masked scalar could be acceptable as long as that value produces an exception if you try to index with it. I would set the data under the mask to be something not otherwise possible, for instance the MAXINT. In practice this probably won't be helping the user much because it will move the exception (from subsequent indexing with a masked scalar) away from the actual problem. But when you start thinking about how to make up arbitrary values, that is a sign of a problem.

mhvk replied:
Unfortunately, I cannot control what `np.ndarray` does with a masked index...  Elsewhere, I just decided not to allow masked indices at all, since it seems very unclear what should happen. But that then suggests the choice is between returning some number or raising an exception on any fully-masked slice. It is tricky... But let's move discussion to #11539...

taldcroft reviewed (commented):
Took another pass through today, with some nitpicky comments. Still need to have a better look at the tests. Getting all the code-cov warnings fixed would be a good thing.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -123,33 +124,39 @@ def _represent_mixin_as_column(col, name, new_cols, mixin_cols,
+        # the format, etc., on the column, but not for ancilliary data (e.g.,
+        # no sense to use a float format for a mask).
+        primary_col = data_attr == col.info._represent_as_dict_primary_data
+        if primary_col:
```
Hmm, I think I used `primary_col` again before realizing that the `info` stripping was in the wrong place below. Anyway, as is, it definitely is only confusing, so reverting to what it was.

mhvk commented on astropy/io/fits/connect.py:
```diff
@@ -111,9 +111,10 @@ def _decode_mixins(tbl):
+    # but not format as that is less flexible in FITS).
     for col in info['datatype']:
-        for attr in ['description', 'meta']:
+        for attr in ['format', 'description', 'meta']:
```
It annoyed me that FITS does not properly serialize formats that one gives, unlike ecsv and hdf5, but you are right it is really a separate issue and here I can just special-case it in the tests.

mhvk replied:
As in fact I had...  I've just removed the relevant commits.

mhvk commented on astropy/io/tests/test_registry.py:
```diff
@@ -508,7 +508,7 @@ class MTable(Table):
         mt = MTable([[1, 2.5]], names=['a'])
         mt['a'].unit = u.m
-        mt['a'].format = '.4f'
+        mt['a'].format = '{:13.4f}'
```
Yes, removed.

mhvk commented:
OK, I really increased coverage today (and of course found bugs...), and renamed the `.unmask()` method to `.filled()`. There are some failures that seem unrelated (-dev and 32 bit do pass).

On the docs, just in case you were like me and had forgotten: the build produces a whole new set one can look at:  https://astropy--11127.org.readthedocs.build/en/11127/utils/masked/index.html
In particular, note that I added some docstrings to some of the numpy function overrides in `function_helpers`, which are included as a module, and give some idea of how masks are interpreted. For the methods, I think the docstrings could do with a bit more work too, but probably best in either the next stage of review or as a follow-up PR.

nstarman commented on astropy/utils/masked/core.py:
```diff
@@ -0,0 +1,1086 @@
+    def filled(self, fill_value=None):
+        """Get the underlying data, possibly filling masked values."""
+        unmasked = self.unmasked
+        if fill_value is None:
+            return unmasked
+        else:
+            unmasked = unmasked.copy()
+            if self.mask.dtype.names:
+                np.ma.core._recursive_filled(unmasked, self.mask, fill_value)
+            else:
+                unmasked[self.mask] = fill_value
+            return unmasked
```
All of 2 lines shorter, but it just makes more evident what is the test case

```suggestion
        if fill_value is not None:
            unmasked = unmasked.copy()
            if self.mask.dtype.names:
                np.ma.core._recursive_filled(unmasked, self.mask, fill_value)
            else:
                unmasked[self.mask] = fill_value
        return unmasked
```

mhvk replied:
Actually, looking at this again, I think really the argument should just be obligatory - if only so that one always gets a copy. But also because unlike `np.ma.MaskedArray`, this class does not carry a fill value around so there is no default. And this has the advantage that one can also fill an object array with `None`.

mhvk replied:
p.s. This is a leftover from when `unmasked = property(filled)`...

nstarman replied:
Sounds good. What's the way to close a suggestion? They always show up for me even when I resolve them.

mhvk replied:
Weird, for me they disappear. Does one have to be author or member perhaps?

nstarman commented on docs/utils/index.rst:
```diff
@@ -26,6 +26,7 @@ The exceptions are below:
 
    iers
    data
+   masked/index
```
Should this be a separate folder, or just a file? Looking at the discussion, there are many followup PRs, e.g. for things like masked SkyCoord in Tables, but currently...

mhvk replied:
I was indeed assuming it would be useful to make this a separate folder. For instance, on how to build your own container class that uses `Masked` etc.

nstarman commented on astropy/utils/shapes.py:
```diff
@@ -194,14 +122,122 @@ def squeeze(self, *args, **kwargs):
+
+        # We cannot just write a generator here, since then the above error
+        # would only be raised once we try to use the iterator, rather than
+        # upon its definition using iter(self).
```
I'm not sure I get this. ``return (self[idx] for idx in range(len(self)))`` doesn't work?

mhvk replied:
I must admit I am unsure why what you write would not work -- it may be quite out of date...  But as I'm just splitting up the class into two pieces here, I think I'd prefer to tackle this in a different PR...

mhvk replied:
Rereading the comment, I think I now remember what it was: if one were to write
```
for idx in range(len(self)):
    yield self[idx]
```
then the check above would not get done.  But your code seems a good alternative. Anyway, for after!

nstarman commented on astropy/utils/shapes.py:
```diff
@@ -194,14 +122,122 @@ def squeeze(self, *args, **kwargs):
+                raise TypeError('scalar {!r} object is not subscriptable.'
+                                .format(self.__class__.__name__))
+            else:
+                raise
```
Cool! I've never tried this. Does this keep the error message?

taldcroft replied:
@nstarman - yup.

nstarman commented on astropy/utils/masked/tests/test_table.py:
```diff
@@ -0,0 +1,111 @@
+
+from .test_masked import assert_masked_equal
+
+try:
```
When rebase, can use the new centralized flags.  #11490

mhvk replied:
OK, done

nstarman commented on astropy/utils/masked/core.py:
```diff
@@ -0,0 +1,1086 @@
+
+        This returns a ``MaskedIterator`` instance, which behaves the same
+        as the `~numpy.flatiter` instance returned by `~numpy.ndarray.flat`,
+        and is similar to, but not a subclass of, Python's built-in iterator
```
is it important that it isn't? MaskedIterator does not currently inherit from anything.

mhvk replied:
Agreed that I can just remove "but not a subclass of" (and probably should do that for `Quantity` too...)

nstarman replied:
Alternatively, would inheriting from python's [``collections.abc.Iterable``](https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable) be beneficial? It guarantees success on ``isinstance(obj, Iterable)`` checks.

mhvk replied:
That `isinstance` check already works (since the class defines an `__iter__()` method.

mhvk replied:
Actually, the iterator differs from the standard python one in that assignment is allowed. I now wrote:
```
        """A 1-D iterator over the Masked array.

        This returns a ``MaskedIterator`` instance, which behaves the same
        as the `~numpy.flatiter` instance returned by `~numpy.ndarray.flat`,
        and is similar to Python's built-in iterator, except that it also
        allows assignment.
        """
```

nstarman reviewed (commented):
This is super  cool and I really like the explanation in the documentation on the shortcomings of the numpy implementation of masking.
I plan on looking at this again, hopefully this coming weekend, when I have some time to check out this branch and run tests, like initializing a CartesianRepresentation with a Masked array...

taldcroft commented:
One issue, making a `Table` from a `MaskedQuantity` loses the mask:
```
ma = Masked([1,2,3], mask=[False, True, False])
mq = Masked([1,2,3]*u.m, mask=[False, True, False])
t = Table([ma, mq], names=['ma', 'mq'])
t
<Table length=3>
  ma     mq  
         m   
int64 float64
----- -------
    1     1.0
    —     2.0
    3     3.0
```
This is of course using what we might call legacy functionality but it is still unexpected. I'd suggest opening a follow-up issue on this.

 What about making a new issue label `utils-masked`?  I think this is going to be big enough to get its own label.

taldcroft commented on astropy/table/serialize.py:
```diff
@@ -123,33 +124,38 @@ def _represent_mixin_as_column(col, name, new_cols, mixin_cols,
     # Store the fully qualified class name
-    obj_attrs['__class__'] = col.__module__ + '.' + col.__class__.__name__
+    obj_attrs.setdefault('__class__',
+                         col.__module__ + '.' + col.__class__.__name__)
```
@nstarman - we might want to look into this as a code simplification, but it's definitely out of scope for this PR.

taldcroft commented on astropy/units/quantity.py:
```diff
@@ -375,7 +375,7 @@ def __new__(cls, value, unit=None, dtype=None, copy=True, order=None,
 
         value = np.array(value, dtype=dtype, copy=copy, order=order,
-                         subok=False, ndmin=ndmin)
+                         subok=True, ndmin=ndmin)
```
@mhvk - this is a notable change in existing code. Can you discuss the implications and why it is OK?

mhvk replied:
I had split this off to #11210 as it is an issue that affects `Distribution` as well, but that is still awaiting proper review.

taldcroft commented on astropy/uncertainty/core.py:
```diff
@@ -281,6 +281,9 @@ def view(self, dtype=None, type=None):
             else:
                 raise ValueError('Cannot set just dtype for a Distribution.')
 
+        if issubclass(type, Distribution):
```
What's the driver here relative to `Masked`?  Is a comment warranted?

mhvk replied:
See #11210 - I'd hoped to rebase once that is in.

taldcroft commented on astropy/utils/masked/core.py:
```diff
@@ -0,0 +1,1086 @@
+            return string
+
+
+class Maskedrecord(np.recarray, MaskedNDArray, data_cls=np.recarray):
```
Should this be named `MaskedRecarray` or `MaskedRecordArray`?

mhvk replied:
Hmm, yes, why was this `record`? I really do not know what I was thinking! The scalar `np.record`? Anyway, I now changed it to `MaskedRecarray`.

taldcroft commented on astropy/utils/masked/tests/test_function_helpers.py:
```diff
@@ -0,0 +1,1294 @@
+More complicated tests of functionality, including with subclasses, are done
+in test_functions.
+
+TODO: finish full coverage.
```
How are things with this TODO?

mhvk replied:
I added a comment:
```
TODO: finish full coverage (see also `~astropy.utils.masked.function_helpers`)
- np.linalg
- np.fft (is there any point?)
- np.lib.nanfunctions
```

taldcroft commented on astropy/utils/masked/tests/test_functions.py:
```diff
@@ -0,0 +1,320 @@
+
+class TestMaskedQuantityConcatenation(TestMaskedArrayConcatenation,
+                                      QuantitySetup):
+    pass
```
Needs completion.

mhvk replied:
This one is OK: it just reruns the tests from `TestMaskedArrayConcatenation` with quantities instead of masked arrays (using `QuantitySetup`).

taldcroft replied:
Ah, clever.

taldcroft reviewed (approved):
I have reviewed this and from my perspective this is OK to merge now, contingent on the liens at the bottom. 

As stated earlier, I haven't done a detailed review of most the new code, but did look at areas which intersect with existing code and don't see any obvious issues. My review was:

- [x] Complete test coverage. I don't see anything from CodeCov. The testing appears extensive.
- [x] Independent review of parts of the code that interact with or impact existing astropy functionality.
- [x] Review of documentation.
- [x] Adherence to standard astropy code / docs practices.
- [x] Skim through all code.

I think there is value in merging this pretty soon and this will make forward progress easier. Most of the code is in entirely new modules, so continued line-by-line review is still possible (even by dropping comments into this PR post-merge). If @nstarman has the time and interest to dive into the code that will be fantastic.

Liens on merging:
- [x] All unresolved comments should be transferred to a new issue.
- [ ] Tests passing.
- [x] @mhvk - commit in principle to supporting this for at least a couple years until it is stable.

As a final point (again) this is really impressive work!

taldcroft commented:
BTW I made a new `utils.masked` label.

pllim commented:
@taldcroft , might also want to update the auto labeling rules to go with the new label. See this for example:

https://github.com/astropy/astropy/blob/ea0426715272f4fc665a5bdf4e0518f6679fe391/.github/labeler.yml#L123-L125

mhvk commented on astropy/coordinates/angles.py:
```diff
@@ -477,7 +477,7 @@ def _no_angle_subclass(obj):
         return tuple(_no_angle_subclass(_obj) for _obj in obj)
 
-    return obj.view(Angle) if isinstance(obj, Angle) else obj
+    return obj.view(Angle) if isinstance(obj, (Latitude, Longitude)) else obj
```
This function is only used inside `Longitude` and `Latitude`, so from that perspective there is no need to change it. Looking at it again, I actually think instead the classes should override `_result_as_quantity`...  I think I'll leave this for follow-up - `angles.py` could use some cleanup generally...

mhvk commented on astropy/utils/masked/core.py:
```diff
@@ -0,0 +1,1086 @@
+        return masked_cls
+
+    @classmethod
+    def _data_mask(cls, data, allow_ma_masked=False):
```
That's fair. I also see I don't describe the `allow_ms_masked` flag (though it is a private method). Do you have a suggestion? Perhaps `_get_unmasked_and_mask`? `_get_data_and_mask` or just `_data_and_mask`?

nstarman replied:
I like ``_get_data_and_mask`` best of the 3 👍

mhvk replied:
OK, I think that is indeed the more pythonesque - `_get_data_and_mask` it is.

mhvk commented on astropy/utils/masked/tests/test_table.py:
```diff
@@ -0,0 +1,111 @@
+
+from .test_masked import assert_masked_equal
+
+try:
```
Done.

mhvk commented:
OK, I think I got all the comments and I made a list for follow up over at #11539.

Main remaining points:
- [ ] Probably best to merge after getting in #11210.
- [ ] Should this be for 4.3 or 5.0? Argument for 5.0 would be that then we have time to get the rest working properly; argument for 4.3 that more people may play around with it, leading to more feedback. I've put it for 4.3 for now.

pllim commented:
> Should this be for 4.3 or 5.0?

If you want people to "beta test" it in 4.3, should be fine to include as long as you make it clear that things are subject to change? We did that to `modeling` for a while back in the days.

nstarman commented:
Does / can / will astropy do alpha and beta releases? Like do a 4.3 stable release and a 4.3+alpha for experimental features?

pllim commented:
> alpha and beta releases?

I don't think we have that precedent. We only do RC and then stable.

mhvk commented:
`-dev` is really where the experimental features live; we don't have explicit releases - I guess like numpy we assume that people test against `-dev`.

mhvk commented on .github/labeler.yml:
```diff
@@ -118,15 +118,17 @@ units:
 utils:
   - cextern/expat/**/*
-  - any: ['**/utils/**/*', '!astropy/utils/iers/**/*', '!docs/utils/iers.rst']
+  - any: ['**/utils/**/*',
```
@pllim - do these additional changes to the labeller make sense? Thanks!

pllim replied:
I think that's correct.

p.s. 100 commits and `+5,427 −112` 😱

mhvk replied:
Yes, this was quite a bit of work. Though most of the lines is somewhat boilerplate in `function_helpers` and lots of tests...

taldcroft commented:
I vote for 4.3 based on precedent from modeling. Essentially all of this is in a new sub-subpackage and it has no impact on existing code. At least that is my understanding. Getting more testing on this from a broad audience will be crucial to ironing out issues and potentially API details.

taldcroft commented:
For 5.0 we should have `Masked` fully integrated with the rest of astropy! No more mixin column table operations exceptions for SkyCoord and Quantity.

mhvk commented:
OK, sounds good! Since #11210 is in, and it seems the rebase was OK too, I'll merge, even though I am sure that if I went over the code again, I'd find some things to improve.  But let that be follow-up that is more reviewable!

Thanks for the reviews, @taldcroft and @nstarman.

pllim commented:
I think this definitely needs a What's New entry. 😉

taldcroft commented:
:tada: :boom:

pllim commented:
I think this broke `pyinstaller` job. See https://github.com/astropy/astropy/runs/2339389527?check_suite_focus=true

mhvk commented:
Darn. But that's an easy fix at least (tells you how long ago this PR started, that I still had relative imports...).
